### ISSUE TITLE
Distributing a source-only version

### ISSUE THREAD

**Issue by ignisf** [toy/image_optim_pack#4]
Hello,

Has it been considered to move away from the 'binaries in git' approach to something like `mini_portile` and binary distributions in rubygems?

See https://github.com/ueno/ruby-gpgme for example?

**Comment by toy** [comment]
The idea of this gem is to make nearly all binaries required for image_optim available without compiling anything as otherwise it is possible to compile everything manually (instructions for some platforms in [image_optim#README](https://github.com/toy/image_optim#binaries-installation)).

Do you suggest it for making binaries available on more platforms or for other reasons?

**Comment by ignisf** [comment]
> Do you suggest it for making binaries available on more platforms or for other reasons?

The benefits would be the following:

 1.  One could download a source distribution that can be compiled on the system if one wishes to or if there is no binary gem version for her platform.
 2.  There won't be any need to have any logic that does architecture detection on runtime, as rubygems will handle this on install of the gem.

As a side note, I'm scratching my own itch here: https://github.com/ignisf/crunch but just wanted to bring it up to your attention.

**Comment by toy** [comment]
1. One can checkout `image_optim_pack` and run `make download`
2. This should not be a problem for platforms on which image_optim_pack will work, the bigger issue is that for those platforms there is a need to download much more than needed, but I will try an idea how to fix this

**Comment by ignisf** [comment]
OK, I now have all the tools except `pngout` which does not seem to be open-source and `svgo`. I consider my issue resolved

**Comment by toy** [comment]
I've tried crunch on a fresh ubuntu64, interesting, but it still requires to install nasm, lippng-dev and libjpeg*-dev

**Comment by ignisf** [comment]
Yes. (I've added this to the README, thanks for reminding me). That's not unlike Nokogiri's deps on `libxml2-dev` and `libxslt-dev`, however those can be easily added to the gem itself in order to have 0 external dependencies.

**Comment by aried3r** [comment]
I'd love to have an official option to compile from source for `image_optim_pack`. It seems to have more visibility than `crunch` at this point and maintaining one repo seems easier.

Of course I can't speak for @ignisf, but he seems interested enough that he'd at least send PRs? Just a thought, I don't want to force anyone to do anything here. :)

**Comment by ignisf** [comment]
If there's will to distribute a source version of image_optim_pack, I'd probably find some time to assist if needed.

@aried3r, keep in mind that I've chosen not to publish ignisf/crunch to rubygems as I don't consider it to be well enough written to be distributed. It's purpose at this point is to scratch a single itch I have in a single project. I could be persuaded to work a bit more on it if I see there's demand but I'd much rather see image_optim_pack get a source-only distribution instead of releasing yet another gem in the wild.

**Comment by toy** [comment]
Could you please describe your idea in a more detail? Should it be a gem which downloads and compiles everything during installation like gpgme or something more like passenger? Should it contain sources or only urls? …

**Comment by ignisf** [comment]
Hey, so I'll use https://rubygems.org/gems/libv8/versions as an example.

First of all, notice how there are binary versions of the gem for each supported platform and a source version for the platforms that there's no binary for.

Whenever you install a source version (https://rubygems.org/gems/libv8/versions/5.0.71.48.3 for example), `ext/libv8/extconf.rb` gets executed which pulls V8's source and compiles it.

Whenever you install a binary version, which rubygems does by default if there's a binary for the current platform (see for example https://rubygems.org/gems/libv8/versions/5.0.71.48.3-x86_64-linux) nothing is compiled. This version of the gem contains only the necessary binaries and no `extconf.rb`. 

The way this is done with the libv8 gem is via two rake tasks:

  * `rake build` compress the gem source and produces a plain source version from the `libv8.gemspec` (the default behaviour of this task)
  * `rake binary` first compiles everything, then cherry-picks which files are needed for the binary version and replaces the file listing in the gemspec before producing a binary version of the gem (see https://github.com/cowboyd/libv8/blob/master/Rakefile#L26-L51).

You could take a look inside the .gem files themselves if I didn't manage to explain well enough.

I was planning to implement a proper `rake binary` task for crunch the other day but got stuck on having all ports produce statically-linked binaries which I wanted to have before stuffing them in an archive to be run on various distros.

**Comment by ignisf** [comment]
As a side note, V8 needs additional arch detection (on top of the one of rubygems) which I expect won't be necessary in this case.

Also, introducing a dependency on `mini_portile2` might not be necessary (and in fact, might be a bad idea) if you have already automated the download and compilation (I only took a quick look at the `Makefile`). The only benefit of it might be the source archive checksum verification.

**Comment by ignisf** [comment]
Basically the issues one would meet with compiling from source are as you mentioned, the dependencies on libpng-dev, libjpeg-dev, zlib and nasm. With the latest few commits in crunch I wanted to make the build self-sufficient and dependent only on the standard library and I also wanted to have the build produce statically-linked executables.

I couldn't quite manage to eliminate the deps yet because of pngquant's ad-hoc build script. I have to rework its "library detection" functions to actually use the paths from the proper environment variables instead of doing ... whatever it is they are doing now.

I haven't started looking at what needs to happen to have the builds produce static executables (and as a matter of fact I'm not sure if it's really necessary if all their deps are available in the binary gem).

**Comment by toy** [comment]
I wanted to have per platform gems in the beginning of creating this gem, but I've found some problems with handling platforms specific gems at that time.
Also I was not sure how to handle jruby, as current solution will work for it, but if using per platform gems, even when on a supported platform, only `ruby` and `jruby` platform specific gems will be chosen. Maybe putting all binaries in `jruby` platform gem can work.
Another also related to `jruby` problem can be caused by bundler as it doesn't allow to choose which platform specific gem to use.

When checking today, rubygems seemed to work fine with platform specific gems, but problems with `jruby` and bundler remain, still it may be worth a try. I'll see how to go with this, as there are few things to do with the gem: cleaning up scripts/make tasks, maybe switching to rake (still not sure).
Also related – there was a PR about adding docker image.

**Comment by ignisf** [comment]
Just FYI, `crunch`'s build should now be self-sufficient and binary gems can be produced with the `rake binary` task.
